plugin/trino-redshift/src/main/java/io/trino/plugin/redshift/RewriteCast.java

```diff
@@ -0,0 +1,92 @@
+import io.trino.spi.connector.ConnectorSession;
+import io.trino.spi.type.BigintType;
+import io.trino.spi.type.DecimalType;
+import io.trino.spi.type.IntegerType;
+import io.trino.spi.type.SmallintType;
+import io.trino.spi.type.TinyintType;
+import io.trino.spi.type.Type;
+
+import java.util.List;
+import java.util.Optional;
+import java.util.function.BiFunction;
+
+import static java.sql.Types.BIGINT;
+import static java.sql.Types.BIT;
+import static java.sql.Types.INTEGER;
+import static java.sql.Types.NUMERIC;
+import static java.sql.Types.SMALLINT;
+import static java.sql.Types.TINYINT;
+
+public class RewriteCast
+        extends AbstractRewriteCast
+{
+    public RewriteCast(BiFunction<ConnectorSession, Type, String> jdbcTypeProvider)
+    {
+        super(jdbcTypeProvider);
+    }
+
+    @Override
+    protected Optional<JdbcTypeHandle> toJdbcTypeHandle(JdbcTypeHandle sourceType, Type targetType)
+    {
+        if (!pushdownSupported(sourceType, targetType)) {
+            return Optional.empty();
+        }
+
+        return switch (targetType) {
+            case TinyintType tinyintType ->
+                    Optional.of(new JdbcTypeHandle(TINYINT, Optional.of(tinyintType.getBaseName()), Optional.empty(), Optional.empty(), Optional.empty(), Optional.empty()));
+            case SmallintType smallintType ->
+                    Optional.of(new JdbcTypeHandle(SMALLINT, Optional.of(smallintType.getBaseName()), Optional.empty(), Optional.empty(), Optional.empty(), Optional.empty()));
+            case IntegerType integerType ->
+                    Optional.of(new JdbcTypeHandle(INTEGER, Optional.of(integerType.getBaseName()), Optional.empty(), Optional.empty(), Optional.empty(), Optional.empty()));
+            case BigintType bigintType ->
+                    Optional.of(new JdbcTypeHandle(BIGINT, Optional.of(bigintType.getBaseName()), Optional.empty(), Optional.empty(), Optional.empty(), Optional.empty()));
+            default -> Optional.empty();
+        };
+    }
+
+    private boolean pushdownSupported(JdbcTypeHandle sourceType, Type targetType)
+    {
+        return switch (targetType) {
+            case TinyintType _, SmallintType _, IntegerType _, BigintType _ ->
+                    List.of(BIT, TINYINT, SMALLINT, INTEGER, BIGINT, NUMERIC).contains(sourceType.jdbcType());
+            default -> false;
+        };
+    }
+
+    @Override
+    protected String buildCast(Type sourceType, Type targetType, String expression, String castType)
+    {
+        if (sourceType instanceof DecimalType && isIntegralType(targetType)) {
+            // Trino rounds up to nearest integral value, whereas Redshift does not.
+            // So using ROUND() to make pushdown same as the trino behavior
+            return "CAST(ROUND(%s) AS %s)".formatted(expression, castType);
+        }
+        return "CAST(%s AS %s)".formatted(expression, castType);
+    }
+
+    private boolean isIntegralType(Type type)
```

Comment:
Isn't this check redundant ?

Reply:
It's redundant as of now since with this PR we are only supporting integral type cast. This is to ensure that in future if we are supporting other types (like `varchar`) then we should not fall under `CAST(ROUND(%s) AS %s)`.